### PR TITLE
Dev/logloss

### PR DESCRIPTION
**Pull request by avolkov-intel**
Add logloss primitive which also can compute hessian and derivative

### CODE REVIEW

File: cpp/oneapi/dal/backend/primitives/blas/gemv.hpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
+/*******************************************************************************
+* Copyright 2022 Intel Corporation
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/blas/gemv.hpp:2]: Please fix year

File: cpp/oneapi/dal/backend/primitives/blas/gemv_dpc.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
+/*******************************************************************************
+* Copyright 2022 Intel Corporation
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/blas/gemv_dpc.cpp:2]: Year

File: cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+
+namespace oneapi::dal::backend::primitives {
+
+template <typename Float>
+sycl::event compute_probabilities(sycl::queue& q,
+                                  const ndview<Float, 1>& parameters,
+                                  const ndview<Float, 2>& data,
+                                  ndview<Float, 1>& probabilities,
+                                  const event_vector& deps) {
+    auto fill_event = fill<Float>(q, probabilities, Float(1), {});
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp:32]: Check that shapes of ndviews are correct and that they have data.

File: cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+        cgh.parallel_for(range, sumReduction, [=](sycl::id<1> idx, auto& sum) {
+            const Float prob = prob_ptr[idx];
+            const std::int32_t label = labels_ptr[idx];
+            sum += -label * sycl::log(prob) - (1 - label) * sycl::log(1 - prob);
+        });
+    });
+
+    auto [out_reg, out_reg_e] = ndarray<Float, 1>::zeros(q, { 1 }, sycl::usm::alloc::device);
+    auto reg_ptr = out_reg.get_mutable_data();
+    event_vector vector_out_reg = { out_reg_e };
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp:100]: What is the purpose to store it as a vector? I see no place where this vector is modified

File: cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+
+    auto param_ptr = parameters.get_data();
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp:102]: Please use `const auto* const` for pointers to constant data and `auto* const` for pointers to modifiable data

File: cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+    auto data_ptr = data.get_data();
+    auto hes_ptr = out_hessian.get_mutable_data();
+    auto proba_ptr = probabilities.get_mutable_data();
+
+    const auto max_wg = device_max_wg_size(q);
+    const auto wg = std::min(p + 1, max_wg);
+    const auto inp_str = data.get_leading_stride();
+    const auto out_str = out_hessian.get_leading_stride();
+
+    const std::int64_t block_size = 32;
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp:345]:
```suggestion
    constexpr std::int64_t block_size = 32;
```

File: cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+    auto hes_event = q.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(deps);
+        cgh.parallel_for(range, [=](sycl::nd_item<3> item) {
+            const std::int64_t obj_ind = item.get_global_id(0);
+            const auto j = item.get_global_id(1);
+            const auto param_ind_2 = item.get_global_id(2);
+            Float val = 0;
+            for (auto k = param_ind_2; k <= j; k += wg) {
+                val = 0;
+                for (auto i = obj_ind * block_size; i < std::min((obj_ind + 1) * block_size, n);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp:358]:
```suggestion
                const auto last_ind = std::min((obj_ind + 1) * block_size, n);
                for (auto i = obj_ind * block_size; i < last_ind;
```

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+            this->p_ = p;
+        }
+        std::cout << "Size: " << n_ << " " << p_ << std::endl;
+
+        const auto dataframe =
+            GENERATE_DATAFRAME(te::dataframe_builder{ n_, p_ }.fill_uniform(-0.5, 0.5));
+        const auto parameters =
+            GENERATE_DATAFRAME(te::dataframe_builder{ 1, p_ + 1 }.fill_uniform(-1, 1));
+        //const auto dataframe = GENERATE_DATAFRAME(te::dataframe_builder{ n_, p_ }.fill_uniform(-10, 10));
+        //const auto parameters = GENERATE_DATAFRAME(te::dataframe_builder{ 1, p_ + 1}.fill_uniform(-2, 2));
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp:60]: Please delete all commented lines

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp [hunk 9]
```diff
@@ -0,0 +1,545 @@
...
+            ndarray<float_t, 1>::empty(this->get_queue(), { n_ }, sycl::usm::alloc::device);
+
+        auto p_event =
+            compute_probabilities(this->get_queue(), params_gpu, data_gpu, out_predictions, {});
+        p_event.wait_and_throw();
+
+        std::chrono::steady_clock::time_point tm2 = std::chrono::steady_clock::now();
+        std::cout << "Compute probabilities: "
+                  << std::chrono::duration_cast<std::chrono::milliseconds>(tm2 - tm1).count()
+                  << "ms" << std::endl;
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp:167]: Please delete "benchmrking" and output

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp [hunk 10]
```diff
@@ -0,0 +1,545 @@
...
+                                                          out_derivative,
+                                                          L1,
+                                                          L2,
+                                                          { fill_event, out_der_e });
+        logloss_event_der.wait_and_throw();
+
+        std::chrono::steady_clock::time_point tm3 = std::chrono::steady_clock::now();
+        std::cout << "Compute logloss and derivative: "
+                  << std::chrono::duration_cast<std::chrono::milliseconds>(tm3 - tm2).count()
+                  << "ms" << std::endl;
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp:190]: The same

**Comment by samir-nasibli** [comment]
I see some conflicts. Please, rebase your branch.

File: cpp/oneapi/dal/backend/primitives/blas/gemv.hpp (hunk 1 above)
Comment (samir-nasibli) [inline cpp/oneapi/dal/backend/primitives/blas/gemv.hpp:2]: 2023

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_perf_dpc.cpp [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+template <typename Param>
+class logloss_perf_test : public te::float_algo_fixture<std::tuple_element_t<0, Param>> {
+public:
+    using float_t = std::tuple_element_t<0, Param>;
+
+    void generate_input(std::int64_t n, std::int64_t p) {
+        this->n_ = n;
+        this->p_ = p;
+        std::cout << "Size: " << n_ << " " << p_ << std::endl;
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_perf_dpc.cpp:47]: output

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_perf_dpc.cpp [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+
+TEMPLATE_TEST_M(logloss_perf_test,
+                "perfomance test small n",
+                "[logloss][100 * 1000]",
+                logloss_types) {
+    this->generate_input(100, 7000);
+    this->measure_time();
+}
+
+} // namespace oneapi::dal::backend::primitives::test
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_perf_dpc.cpp:157]: Please fix EOF

**Review by KulikovNikita — CHANGES_REQUESTED** [review]

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_perf_dpc.cpp [hunk 13]
```diff
@@ -0,0 +1,143 @@
+/*******************************************************************************
+* Copyright 2022 Intel Corporation
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_perf_dpc.cpp:2]: Year

File: cpp/oneapi/dal/backend/primitives/reduction/reduction_1d.hpp [hunk 14]
```diff
@@ -0,0 +1,101 @@
+/*******************************************************************************
+* Copyright 2022 Intel Corporation
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/reduction/reduction_1d.hpp:2]: Year

File: cpp/oneapi/dal/backend/primitives/reduction/reduction_1d_dpc.cpp [hunk 15]
```diff
@@ -0,0 +1,97 @@
+/*******************************************************************************
+* Copyright 2022 Intel Corporation
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/reduction/reduction_1d_dpc.cpp:2]: Year

File: cpp/oneapi/dal/backend/primitives/reduction/test/reduction_1d_dpc.cpp [hunk 16]
```diff
@@ -0,0 +1,106 @@
+/*******************************************************************************
+* Copyright 2022 Intel Corporation
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/reduction/test/reduction_1d_dpc.cpp:2]: Year

File: cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp [hunk 17]
```diff
@@ -0,0 +1,471 @@
...
+        cgh.parallel_for(range, sumReduction, [=](sycl::id<1> idx, auto& sum) {
+            const Float prob = prob_ptr[idx];
+            const std::int32_t label = labels_ptr[idx];
+            sum += -label * sycl::log(prob) - (1 - label) * sycl::log(1 - prob);
+        });
+    });
+
+    auto [out_reg, out_reg_e] = ndarray<Float, 1>::zeros(q, { 1 }, sycl::usm::alloc::device);
+    auto* const reg_ptr = out_reg.get_mutable_data();
+    event_vector vector_out_reg = { out_reg_e };
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp:104]:
```suggestion
    const event_vector vector_out_reg = { out_reg_e };
```

File: cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp [hunk 18]
```diff
@@ -0,0 +1,471 @@
...
+            const Float param = param_ptr[idx];
+            out_derivative_ptr[idx] += sycl::copysign(L1, param) + L2 * 2 * param;
+        });
+    });
+
+    return reg_event;
+}
+
+template <typename Float>
+
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp:334]: empty line

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp [hunk 19]
```diff
@@ -0,0 +1,402 @@
...
+        std::srand(2007 + n_);
+        auto ptr_lab = this->labels_.get_mutable_data();
+        for (auto i = 0; i < n_; ++i) {
+            ptr_lab[i] = std::rand() % 2;
+        }
+    }
+
+    void run_test() {
+        float_t L1 = 1.2;
+        float_t L2 = 0.7;
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp:80]:
```suggestion
        constexpr float_t L1 = 1.2;
        constexpr float_t L2 = 0.7;
```

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp [hunk 20]
```diff
@@ -0,0 +1,402 @@
...
+        constexpr std::int64_t p = 3;
+        const float_t data[n * p] = {
+            0.83731708, -0.70899924, -1.23362082, 0.23468538, -0.10549413,
+            1.12902673, -0.61035703, -1.55617932, 0.60419908, 0.30589827,
+            0.63919892, -0.23380754, 2.38196927,  1.64158111, 0.13677077
+        };
+        const std::int32_t labels[n] = { 0, 1, 1, 0, 1 };
+        const float_t L1 = 2.3456;
+        const float_t L2 = 3.123;
+        const float_t cur_param[p + 1] = { -0.2, 0.1, -1, 0.4 };
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp:103]:
```suggestion
        constexpr float_t data[n * p] = {
            0.83731708, -0.70899924, -1.23362082, 0.23468538, -0.10549413,
            1.12902673, -0.61035703, -1.55617932, 0.60419908, 0.30589827,
            0.63919892, -0.23380754, 2.38196927,  1.64158111, 0.13677077
        };
        constexpr std::int32_t labels[n] = { 0, 1, 1, 0, 1 };
        constexpr float_t L1 = 2.3456;
        constexpr float_t L2 = 3.123;
        constexpr float_t cur_param[p + 1] = { -0.2, 0.1, -1, 0.4 };
```

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp [hunk 21]
```diff
@@ -0,0 +1,402 @@
...
+    }
+
+    void test_input(const ndarray<float_t, 2>& data_host,
+                    const ndarray<float_t, 1>& params_host,
+                    const ndarray<std::int32_t, 1>& labels_host,
+                    const float_t L1,
+                    const float_t L2) {
+        constexpr float_t eps = sizeof(float_t) >= 8 ? 1e-6 : 1e-4; //1e-4;
+        std::int64_t n = data_host.get_dimension(0);
+        std::int64_t p = data_host.get_dimension(1);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp:121]:
```suggestion
        const std::int64_t n = data_host.get_dimension(0);
        const std::int64_t p = data_host.get_dimension(1);
```

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp [hunk 22]
```diff
@@ -0,0 +1,402 @@
...
+
+        SUCCEED();
+    }
+
+    void test_input(const ndarray<float_t, 2>& data_host,
+                    const ndarray<float_t, 1>& params_host,
+                    const ndarray<std::int32_t, 1>& labels_host,
+                    const float_t L1,
+                    const float_t L2) {
+        constexpr float_t eps = sizeof(float_t) >= 8 ? 1e-6 : 1e-4; //1e-4;
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp:119]:
```suggestion
        constexpr float_t rtol = sizeof(float_t) > 4 ? 1e-6 : 1e-4;
        constexpr float_t atol = sizeof(float_t) > 4 ? 1e-6 : 1e-4;
```

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp [hunk 23]
```diff
@@ -0,0 +1,402 @@
...
+                                                          out_predictions,
+                                                          out_logloss,
+                                                          out_derivative,
+                                                          L1,
+                                                          L2,
+                                                          { fill_event, out_der_e });
+        logloss_event_der.wait_and_throw();
+        auto out_derivative_host = out_derivative.to_host(this->get_queue());
+        val_logloss = out_logloss.to_host(this->get_queue(), {}).at(0);
+        check_val(val_logloss, logloss, eps);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp:175]:
```suggestion
        check_val(val_logloss, logloss, rtol, atol);
```

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp [hunk 24]
```diff
@@ -0,0 +1,402 @@
...
+        sycl::event logloss_event = compute_logloss(this->get_queue(),
+                                                    params_gpu,
+                                                    data_gpu,
+                                                    labels_gpu,
+                                                    out_logloss,
+                                                    L1,
+                                                    L2,
+                                                    { out_e });
+        logloss_event.wait_and_throw();
+        float_t val_logloss = out_logloss.to_host(this->get_queue(), {}).at(0);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp:155]:
```suggestion
        const float_t val_logloss1 = out_logloss.to_host(this->get_queue(), {}).at(0);
```

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp [hunk 25]
```diff
@@ -0,0 +1,402 @@
...
+                                                          labels_gpu,
+                                                          out_predictions,
+                                                          out_logloss,
+                                                          out_derivative,
+                                                          L1,
+                                                          L2,
+                                                          { fill_event, out_der_e });
+        logloss_event_der.wait_and_throw();
+        auto out_derivative_host = out_derivative.to_host(this->get_queue());
+        val_logloss = out_logloss.to_host(this->get_queue(), {}).at(0);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp:174]:
```suggestion
        const auto val_logloss2 = out_logloss.to_host(this->get_queue(), {}).at(0);
```

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp [hunk 26]
```diff
@@ -0,0 +1,402 @@
...
+        auto data_host = ndarray<float_t, 2>::wrap(data, { n, p });
+        auto labels_host = ndarray<std::int32_t, 1>::wrap(labels, n);
+        auto params_host = ndarray<float_t, 1>::wrap(cur_param, p + 1);
+
+        test_input(data_host, params_host, labels_host, L1, L2);
+
+        SUCCEED();
+    }
+
+    void test_input(const ndarray<float_t, 2>& data_host,
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp:114]: Try to split into the smaller functions

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp [hunk 27]
```diff
@@ -0,0 +1,402 @@
...
+        auto out_predictions =
+            ndarray<float_t, 1>::empty(this->get_queue(), { n }, sycl::usm::alloc::device);
+
+        auto p_event =
+            compute_probabilities(this->get_queue(), params_gpu, data_gpu, out_predictions, {});
+        p_event.wait_and_throw();
+
+        auto predictions_host = out_predictions.to_host(this->get_queue(), {});
+
+        float_t logloss = test_predictions_and_logloss(data_host,
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp:136]:
```suggestion
        const float_t logloss = test_predictions_and_logloss(data_host,
```

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp [hunk 28]
```diff
@@ -0,0 +1,402 @@
...
+        if (sizeof(logloss) >= 4) {
+            test_derivative_and_hessian(data_gpu,
+                                        labels_gpu,
+                                        out_derivative_host,
+                                        hessian_host,
+                                        params_host,
+                                        L1,
+                                        L2,
+                                        eps * 100);
+        }
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp:218]: Please check

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_perf_dpc.cpp [hunk 29]
```diff
@@ -0,0 +1,143 @@
...
+        std::srand(2007 + n_);
+        auto ptr_lab = this->labels_.get_mutable_data();
+        for (auto i = 0; i < n_; ++i) {
+            ptr_lab[i] = std::rand() % 2;
+        }
+    }
+
+    void measure_time() {
+        float_t L1 = 1.2;
+        float_t L2 = 0.7;
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_perf_dpc.cpp:64]:
```suggestion
        constexpr float_t L1 = 1.2;
        constexpr float_t L2 = 0.7;
```

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_perf_dpc.cpp [hunk 30]
```diff
@@ -0,0 +1,143 @@
...
+            auto logloss_event_der = compute_logloss_with_der(this->get_queue(),
+                                                              params_gpu,
+                                                              data_gpu,
+                                                              labels_gpu,
+                                                              out_predictions,
+                                                              out_logloss,
+                                                              out_derivative,
+                                                              L1,
+                                                              L2,
+                                                              { fill_event, out_der_e });
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_perf_dpc.cpp:99]: You need to measure perf only for this section

File: cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp (hunk 17 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp:104]: Still valid

File: cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp (hunk 18 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp:334]: Still valid

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp [hunk 31]
```diff
@@ -0,0 +1,498 @@
...
+                          ndarray<long double, 1>& out_der,
+                          float_t L1,
+                          float_t L2) {
+        const std::int64_t n = data.get_dimension(0);
+        const std::int64_t p = data.get_dimension(1);
+        for (std::int64_t j = 0; j <= p; ++j) {
+            long double val = 0;
+            for (std::int64_t i = 0; i < n; ++i) {
+                long double x1 = j > 0 ? data.at(i, j - 1) : 1;
+                long double prob = probabilities.at(i);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp:298]: As far as I know `long double` == `double` for the absolute majority of systems

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp [hunk 32]
```diff
@@ -0,0 +1,498 @@
...
+                                     const float_t atol = 1e-3) {
+        const std::int64_t n = data.get_dimension(0);
+        const std::int64_t p = data.get_dimension(1);
+        constexpr std::int64_t MAXN = 2000;
+        constexpr float_t step = sizeof(float_t) > 4 ? 1e-4 : 1e-3;
+
+        const auto data_host = data.to_host(this->get_queue());
+        const auto labels_host = labels.to_host(this->get_queue());
+
+        float_t cur_param[MAXN];
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp:385]: Please use safer alternative:

```suggestion
        std::array<float_t, max_n> cur_param;
```

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp [hunk 33]
```diff
@@ -0,0 +1,498 @@
...
+                                     const ndarray<float_t, 1>& derivative,
+                                     const ndarray<float_t, 2>& hessian,
+                                     const ndarray<float_t, 1>& params_host,
+                                     const float_t L1,
+                                     const float_t L2,
+                                     const float_t rtol = 1e-3,
+                                     const float_t atol = 1e-3) {
+        const std::int64_t n = data.get_dimension(0);
+        const std::int64_t p = data.get_dimension(1);
+        constexpr std::int64_t MAXN = 2000;
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp:379]:
```suggestion
        constexpr std::int64_t max_n = 2000;
```

**Review by KulikovNikita — APPROVED** [review]
Minor fixes required. Please take and fix

**Comment by KulikovNikita** [comment]
I believe you can merge it after passing CI

**Pull request merged**